# Combining Datasets


In [1]:
import pandas as pd
import numpy as np

wdi_df = pd.read_csv("./data/wdi_clean.csv")

wdi_df.head()


Country Name Country Code  \
0  European Union (27)          EUU   
1  European Union (27)          EUU   
2  European Union (27)          EUU   
3  European Union (27)          EUU   
4  European Union (27)          EUU   

                                      Indicator Name     Indicator Code  \
0  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS   
1  Access to clean fuels and technologies for coo...  EG.CFT.ACCS.RU.ZS   
2  Access to clean fuels and technologies for coo...  EG.CFT.ACCS.UR.ZS   
3            Access to electricity (% of population)     EG.ELC.ACCS.ZS   
4  Access to electricity, rural (% of rural popul...  EG.ELC.ACCS.RU.ZS   

    1990   1991   1992   1993   1994   1995  ...        2010        2011  \
0    NaN    NaN    NaN    NaN    NaN    NaN  ...   98.995793   99.060575   
1    NaN    NaN    NaN    NaN    NaN    NaN  ...   97.497321   97.638137   
2    NaN    NaN    NaN    NaN    NaN    NaN  ...   99.869291   99.869739   
3  100.0  100.0  100.0  100.0  100.0  100.0  ...  100.000000  100.000000   
4  100.0  100.0  100.0  100.0  100.0  100.0  ...  100.000000  100.000000   

         2012        2013        2014        2015        2016        2017  \
0   99.098573   99.134294   99.169027   99.197882   99.221345   99.424531   
1   97.732486   97.857445   97.967160   98.018315   98.103910   98.208828   
2   99.874219   99.878765   99.879905   99.881077   99.882307   99.883422   
3  100.000000  100.000000  100.000000  100.000000   99.960196  100.000000   
4  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   

         2018        2019  
0   99.428916   99.458748  
1   98.274355   98.321068  
2   99.881291   99.888532  
3  100.000000  100.000000  
4  100.000000  100.000000  

[5 rows x 34 columns]

In [2]:
climate_df = pd.read_csv("./data/climate_watch_data.csv")

climate_df.head()


Country Data source                Sector      Gas    Unit      2019  \
0   World        CAIT  Total including LUCF  All GHG  MtCO₂e  49758.23   
1   World        CAIT  Total excluding LUCF  All GHG  MtCO₂e  48116.56   
2   World        CAIT                Energy  All GHG  MtCO₂e  37636.10   
3   World        CAIT  Total including LUCF      CO2  MtCO₂e  36874.11   
4   World        CAIT  Total excluding LUCF      CO2  MtCO₂e  35512.86   

       2018      2017      2016      2015  ...      1999      1998      1997  \
0  49368.04  48251.88  47531.68  46871.77  ...  35101.90  35099.21  35537.18   
1  47980.47  47031.82  46264.07  46085.31  ...  33282.05  33088.07  32855.85   
2  37603.22  36777.63  36188.89  36173.70  ...  25389.87  25257.77  25075.28   
3  36669.40  35736.22  35223.63  34559.71  ...  25064.85  25003.46  25396.41   
4  35476.59  34689.96  34144.04  34076.29  ...  23506.53  23348.62  23185.61   

       1996      1995      1994      1993      1992      1991      1990  
0  34179.33  33805.61  33015.04  32729.06  32588.09  32670.51  32523.58  
1  32467.96  31890.42  31105.68  30819.73  30678.77  30761.18  30614.25  
2  24656.11  24150.88  23501.77  23395.58  23265.41  23365.41  23244.24  
3  24289.64  23964.26  23334.16  23191.71  23049.37  23098.30  22943.37  
4  22769.14  22289.91  21665.65  21523.23  21380.89  21429.81  21274.89  

[5 rows x 35 columns]

In [3]:
un_sds_df = pd.read_csv("./data/un_sds_clean.csv")

un_sds_df.head()


SeriesCode                                  SeriesDescription  \
0  DC_ODA_POVDLG  Official development assistance grants for pov...   
1  DC_ODA_POVDLG  Official development assistance grants for pov...   
2  DC_ODA_POVDLG  Official development assistance grants for pov...   
3  DC_ODA_POVDLG  Official development assistance grants for pov...   
4  DC_ODA_POVDLG  Official development assistance grants for pov...   

  GeoAreaName  Age Location  Sex    Units  1990  1991  1992  ...    2010  \
0   Australia  NaN      NaN  NaN  PERCENT   NaN   NaN   NaN  ...  0.0329   
1     Austria  NaN      NaN  NaN  PERCENT   NaN   NaN   NaN  ...  0.0071   
2  Azerbaijan  NaN      NaN  NaN  PERCENT   NaN   NaN   NaN  ...     NaN   
3     Belgium  NaN      NaN  NaN  PERCENT   NaN   NaN   NaN  ...  0.0345   
4      Canada  NaN      NaN  NaN  PERCENT   NaN   NaN   NaN  ...  0.0364   

     2011    2012    2013    2014    2015    2016    2017    2018    2019  
0  0.0376  0.0383  0.0321  0.0307  0.0325  0.0230  0.0235  0.0296  0.0220  
1  0.0066  0.0073  0.0062  0.0086  0.0057  0.0053  0.0045  0.0056  0.0046  
2     NaN     NaN     NaN  0.0013     NaN  0.0004     NaN  0.0002  0.0047  
3  0.0315  0.0274  0.0234  0.0265  0.0215  0.0233  0.0199  0.0187  0.0191  
4  0.0410  0.0409  0.0422  0.0317  0.0301  0.0303  0.0277  0.0261  0.0217  

[5 rows x 37 columns]

# Put Datasets into Common Column Configuration

| Column     | Description                                                                |
| ---------- | -------------------------------------------------------------------------- |
| country    | Country name                                                               |
| c_code     | Three letter shortcode for the country                                     |
| indicator  | Name for the value, for GHG data this describes the source of the emission |
| i_code     | Shorthand code for the indicator                                           |
| gas        | Type of emission gas (“CO2", “CH4”, etc.)                                  |
| 1990..2019 | Range of years for the data                                                |

Generating the list of years that we are focusing on for our analysis, 1990-2019.


In [4]:
years = list(map(str, range(1990, 2020)))


In [5]:
wdi_df_renamed = wdi_df.rename(
    {
        "Country Name": "country",
        "Country Code": "c_code",
        "Indicator Name": "indicator",
        "Indicator Code": "i_code",
    },
    axis="columns",
)

wdi_df_renamed.insert(4, "gas", "")

wdi_df_renamed.head()


country c_code  \
0  European Union (27)    EUU   
1  European Union (27)    EUU   
2  European Union (27)    EUU   
3  European Union (27)    EUU   
4  European Union (27)    EUU   

                                           indicator             i_code gas  \
0  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS       
1  Access to clean fuels and technologies for coo...  EG.CFT.ACCS.RU.ZS       
2  Access to clean fuels and technologies for coo...  EG.CFT.ACCS.UR.ZS       
3            Access to electricity (% of population)     EG.ELC.ACCS.ZS       
4  Access to electricity, rural (% of rural popul...  EG.ELC.ACCS.RU.ZS       

    1990   1991   1992   1993   1994  ...        2010        2011        2012  \
0    NaN    NaN    NaN    NaN    NaN  ...   98.995793   99.060575   99.098573   
1    NaN    NaN    NaN    NaN    NaN  ...   97.497321   97.638137   97.732486   
2    NaN    NaN    NaN    NaN    NaN  ...   99.869291   99.869739   99.874219   
3  100.0  100.0  100.0  100.0  100.0  ...  100.000000  100.000000  100.000000   
4  100.0  100.0  100.0  100.0  100.0  ...  100.000000  100.000000  100.000000   

         2013        2014        2015        2016        2017        2018  \
0   99.134294   99.169027   99.197882   99.221345   99.424531   99.428916   
1   97.857445   97.967160   98.018315   98.103910   98.208828   98.274355   
2   99.878765   99.879905   99.881077   99.882307   99.883422   99.881291   
3  100.000000  100.000000  100.000000   99.960196  100.000000  100.000000   
4  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   

         2019  
0   99.458748  
1   98.321068  
2   99.888532  
3  100.000000  
4  100.000000  

[5 rows x 35 columns]

In [6]:
codes = dict(zip(wdi_df_renamed["country"].unique(), wdi_df_renamed["c_code"].unique()))


def set_country_code(row):
    try:
        row["c_code"] = codes[row["country"]]
    except:
        pass
    return row


In [7]:
un_sds_df["indicator"] = (
    un_sds_df["SeriesDescription"]
    + un_sds_df["Units"].astype(str)
    + un_sds_df["Age"].replace({np.NaN: ""})
    + un_sds_df["Location"].replace({np.NaN: ""})
    + un_sds_df["Sex"].replace({np.NaN: ""})
)


un_sds_df_renamed = un_sds_df.rename(
    {"GeoAreaName": "country", "SeriesCode": "i_code"},
    axis="columns",
)

un_sds_df_renamed = un_sds_df_renamed.apply(set_country_code, axis="columns").reindex(
    [
        "country",
        "c_code",
        "indicator",
        "i_code",
        *years,
    ],
    axis="columns",
)


un_sds_df_renamed.insert(4, "gas", "")

un_sds_df_renamed.head()


country c_code                                          indicator  \
0   Australia    AUS  Official development assistance grants for pov...   
1     Austria    AUT  Official development assistance grants for pov...   
2  Azerbaijan    AZE  Official development assistance grants for pov...   
3     Belgium    BEL  Official development assistance grants for pov...   
4      Canada    CAN  Official development assistance grants for pov...   

          i_code gas  1990  1991  1992  1993  1994  ...    2010    2011  \
0  DC_ODA_POVDLG       NaN   NaN   NaN   NaN   NaN  ...  0.0329  0.0376   
1  DC_ODA_POVDLG       NaN   NaN   NaN   NaN   NaN  ...  0.0071  0.0066   
2  DC_ODA_POVDLG       NaN   NaN   NaN   NaN   NaN  ...     NaN     NaN   
3  DC_ODA_POVDLG       NaN   NaN   NaN   NaN   NaN  ...  0.0345  0.0315   
4  DC_ODA_POVDLG       NaN   NaN   NaN   NaN   NaN  ...  0.0364  0.0410   

     2012    2013    2014    2015    2016    2017    2018    2019  
0  0.0383  0.0321  0.0307  0.0325  0.0230  0.0235  0.0296  0.0220  
1  0.0073  0.0062  0.0086  0.0057  0.0053  0.0045  0.0056  0.0046  
2     NaN     NaN  0.0013     NaN  0.0004     NaN  0.0002  0.0047  
3  0.0274  0.0234  0.0265  0.0215  0.0233  0.0199  0.0187  0.0191  
4  0.0409  0.0422  0.0317  0.0301  0.0303  0.0277  0.0261  0.0217  

[5 rows x 35 columns]

In [8]:
climate_df_renamed = climate_df.drop(["Data source", "Unit"], axis="columns").rename(
    {
        "Country": "country",
        "Sector": "indicator",
        "Gas": "gas",
    },
    axis="columns",
)

climate_df_renamed = climate_df_renamed.apply(set_country_code, axis="columns").reindex(
    ["country", "c_code", "indicator", "gas", *years], axis="columns"
)

climate_df_renamed.insert(3, "i_code", "")

climate_df_renamed.head()


country c_code             indicator i_code      gas      1990      1991  \
0   World    WLD  Total including LUCF         All GHG  32523.58  32670.51   
1   World    WLD  Total excluding LUCF         All GHG  30614.25  30761.18   
2   World    WLD                Energy         All GHG  23244.24  23365.41   
3   World    WLD  Total including LUCF             CO2  22943.37  23098.30   
4   World    WLD  Total excluding LUCF             CO2  21274.89  21429.81   

       1992      1993      1994  ...      2010      2011      2012      2013  \
0  32588.09  32729.06  33015.04  ...  44877.51  45041.81  45597.14  46238.61   
1  30678.77  30819.73  31105.68  ...  43387.31  44637.13  45164.50  45850.35   
2  23265.41  23395.58  23501.77  ...  34193.57  35154.37  35563.41  36184.99   
3  23049.37  23191.71  23334.16  ...  33359.98  33264.47  33657.18  34302.12   
4  21380.89  21523.23  21665.65  ...  32084.17  33063.81  33445.78  34107.98   

       2014      2015      2016      2017      2018      2019  
0  46881.78  46871.77  47531.68  48251.88  49368.04  49758.23  
1  46143.84  46085.31  46264.07  47031.82  47980.47  48116.56  
2  36279.47  36173.70  36188.89  36777.63  37603.22  37636.10  
3  34689.40  34559.71  35223.63  35736.22  36669.40  36874.11  
4  34229.56  34076.29  34144.04  34689.96  35476.59  35512.86  

[5 rows x 35 columns]

For the WDI data we are filtering this down to only GDP data which is relevant to our analysis.


In [9]:
wdi_df_gdp = wdi_df_renamed[wdi_df_renamed["indicator"].str.contains("gdp", case=False)]

combined_df = pd.concat(
    [climate_df_renamed, un_sds_df_renamed, wdi_df_gdp]
).sort_values(["country", "indicator", "gas"])

combined_df.head()


country c_code                                          indicator  \
1813  Afghanistan    AFG                                        Agriculture   
2144  Afghanistan    AFG                                        Agriculture   
3017  Afghanistan    AFG                                        Agriculture   
1840  Afghanistan    AFG  Agriculture, forestry, and fishing, value adde...   
187   Afghanistan    AFG  Average proportion of deprivations for people ...   

              i_code      gas  1990  1991  1992  1993  1994  ...       2010  \
1813                  All GHG  8.07  8.40  8.41  8.49  8.52  ...  15.790000   
2144                      CH4  5.36  5.61  5.67  5.72  5.93  ...  11.510000   
3017                      N2O  2.71  2.79  2.74  2.77  2.60  ...   4.270000   
1840  NV.AGR.TOTL.ZS            NaN   NaN   NaN   NaN   NaN  ...  26.210069   
187      SD_MDP_ANDI            NaN   NaN   NaN   NaN   NaN  ...        NaN   

           2011       2012       2013       2014       2015       2016  \
1813  15.900000  15.780000  15.720000  16.220000  15.110000  15.320000   
2144  11.530000  11.380000  11.280000  11.480000  10.850000  10.630000   
3017   4.370000   4.400000   4.440000   4.740000   4.260000   4.680000   
1840  23.743664  24.390874  22.810663  22.137041  20.634323  25.740314   
187         NaN        NaN        NaN        NaN        NaN  52.500000   

           2017       2018       2019  
1813  15.220000  14.850000  15.540000  
2144  10.330000  10.540000  10.920000  
3017   4.900000   4.310000   4.630000  
1840  26.420199  22.042897  25.773971  
187         NaN        NaN        NaN  

[5 rows x 35 columns]

In [10]:
# len(combined_df["indicator"].unique())
# len(combined_df["country"].unique())
combined_df.columns


Index(['country', 'c_code', 'indicator', 'i_code', 'gas', '1990', '1991',
       '1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000',
       '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009',
       '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017', '2018',
       '2019'],
      dtype='object')

Our combined dataset contains `34,667` rows and is about `10.9 MB` as a csv.


In [11]:
# Uncomment to replace the complete_dataset
# combined_df.to_csv("./data/complete_dataset.csv", index=False)

# Output the first 100 records as a sample set
combined_df.head(100).to_csv("./data/complete_dataset_sample.csv", index=False)


In [12]:
%reload_ext watermark

%watermark -iv -v -m

Python implementation: CPython
Python version       : 3.10.6
IPython version      : 8.5.0

Compiler    : Clang 13.1.6 (clang-1316.0.21.2.5)
OS          : Darwin
Release     : 21.5.0
Machine     : x86_64
Processor   : i386
CPU cores   : 8
Architecture: 64bit

numpy : 1.23.3
pandas: 1.5.0
sys   : 3.10.6 (main, Aug 30 2022, 05:12:36) [Clang 13.1.6 (clang-1316.0.21.2.5)]

